### PR TITLE
seslib: remove host's virtual networks on destroy

### PR DESCRIPTION
**Pull request by jecluis**
`Signed-off-by: Joao Eduardo Luis <joao@suse.com>`

### CODE REVIEW

**Comment by jecluis** [comment]
This ought to address https://github.com/SUSE/sesdev/issues/17

**Comment by jecluis** [comment]
huh... from https://github.com/SUSE/sesdev/pull/39 it seems I may be risking destroying pre-existing networks. I guess adding a "--destroy-networks" to `sesdev destroy` should address that; or do something smarter, like ensuring there's nothing else using that network? The latter seems a bit too risky too though. The former might be the safest option.

**Comment by smithfarm** [comment]
@jecluis Are you planning to look at the linter complaints?

**Comment by jschmid1** [comment]
As @jecluis pointed out, this will remove networks that may be used by other `domains`. It appears that `network.destroy` will not check if there are any references[0] to it.

To me it sounds like we should implement a check for references in order to make this less scary..

[0] https://libvirt.org/html/libvirt-libvirt-network.html#virNetworkDestroy (python bindings are a 1:1 mapping to the C api)

**Comment by jschmid1** [comment]
in addition to that I just noticed that a network needs to be `active` to get destroyed (for whatever reason..)

```Error while removing network bsc_11601722. Call to virNetworkDestroy failed: Requested operation is not valid: network 'bsc_11601722' is not active.```

**Comment by jecluis** [comment]
@ all, yeah, will go back to this later this week, I intend to add a `--destroy-networks` option, fix the linting issues (which, in my not-so-much-a-defense, I kinda didn't realize was a thing), and will try to reproduce @jschmid1's issue.

**Comment by jecluis** [comment]
here's the thing... after upgrading to vagrant 2.2.6 and vagrant-libvirt 0.0.45, seems like the networks get automatically removed upon destroy. Is anyone able to verify that?

**Comment by jan--f** [comment]
> here's the thing... after upgrading to vagrant 2.2.6 and vagrant-libvirt 0.0.45, seems like the networks get automatically removed upon destroy. Is anyone able to verify that?

Can't confirm this, been running this setup for a while now, though with a vagrant-libvirt build from master (still 0.0.45 though).

**Comment by jecluis** [comment]
Well, that is annoying then... _sigh_

**Comment by jecluis** [comment]
I can't, for the life of me, reproduce this anymore.

> (venv) ➜ joao@tagide sesdev git:(master) virsh --connect qemu:///system net-list --all
>  Name              State      Autostart   Persistent
> ------------------------------------------------------
>  default           inactive   no          yes
>  octopus_test0     inactive   no          yes
>  octopus_test20    active     no          yes
>  vagrant-libvirt   active     no          yes
> 
> (venv) ➜ joao@tagide sesdev git:(master) sesdev destroy octopus_test
> Are you sure you want to destroy the cluster? [y/N]: y
> === Running shell command ===
> vagrant destroy admin --force
> ==> admin: Removing domain...
> Command '['vagrant', 'destroy', 'admin', '--force']' failed: ret=1 stderr:
> b"Error while removing network octopus_test0. Call to virNetworkDestroy failed: Requested operation is not valid: network 'octopus_test0' is not active.\n"
>
> (venv) ➜ joao@tagide sesdev git:(master) sesdev destroy octopus_test2 
> Are you sure you want to destroy the cluster? [y/N]: y
> === Running shell command ===
> vagrant destroy admin --force
> ==> admin: Removing domain...
>
> (venv) ➜ joao@tagide sesdev git:(master) virsh --connect qemu:///system net-list --all
>  Name              State      Autostart   Persistent
> ------------------------------------------------------
>  default           inactive   no          yes
>  octopus_test0     inactive   no          yes

And this is vanilla sesdev at `1b42b20`. The only thing that was easily reproducible was @jschmid1 issue with the inactive network.

Can someone provide the steps they are using to reproduce this? I used to be able to consistently reproduce this simply by creating a sesdev cluster and then destroying it, but now... sad-panda.png

**Comment by smithfarm** [comment]
@jecluis I am running

```
(venv) smithfarm@geras:~/sesdev> rpm -q vagrant
vagrant-2.2.7-3.3.x86_64
(venv) smithfarm@geras:~/sesdev> rpm -q vagrant-libvirt
vagrant-libvirt-0.0.45-17.20.x86_64
```

and it doesn't look like the networks are getting auto-destroyed. For example:

```
(venv) smithfarm@geras:~/sesdev> sesdev list
| Deployments |  Version   |     Status      |                             VMs                              |
-------------------------------------------------------------------------------------------------------------
| octopus     | octopus    | running         | admin                                                        |
| ses7        | ses7       | running         | admin    
```

Yet:

```
(venv) smithfarm@geras:~/sesdev> sudo virsh net-list
 Name              State    Autostart   Persistent
----------------------------------------------------
 octopus0          active   yes         yes
 octopus1          active   yes         yes
 octopus10         active   yes         yes
 octopus2          active   yes         yes
 octopus3          active   yes         yes
 octopus4          active   yes         yes
 octopus5          active   yes         yes
 octopus6          active   yes         yes
 octopus7          active   yes         yes
 octopus8          active   yes         yes
 octopus9          active   yes         yes
 ses70             active   yes         yes
 vagrant-libvirt   active   no          yes
```

**Comment by smithfarm** [comment]
@jecluis Also, do I understand correctly that you are proposing to implement a `--destroy-networks` option to `sesdev destroy` which would cause any left-over networks to be removed (undefined/destroyed) by name?

In other words, in the case above, where there are a bunch of dangling `octopus*` networks, if I do `vagrant destroy --destroy-networks octopus` all of those networks will disappear?

That would be awesome!

File: seslib/__init__.py [hunk 1]
```diff
@@ -502,6 +529,24 @@ def remove_image(self, image_name):
...
+        try:
+            network = self.libvirt_conn.networkLookupByName(name)
+        except libvirt.libvirtError:
+            logger.warning("Unable to find network '%s' for removal", name)
+            return False
+
+        try:
+            network.destroy()
+        except libvirt.libvirtError:
+            logger.error("Something went wrong destroying network '%s'", name)
```
Comment (smithfarm) [inline seslib/__init__.py]: do we need a `return False` here in this `except` block?

As it is, if this code path is triggered, won't we see the following in the log?

```
ERROR: Something went wrong destroying network 'foo'
INFO: Destroyed network 'foo'
```

?

Comment (jecluis) [reply seslib/__init__.py]: oops. Yeah, for some reason that was not staged. _sigh_

**Comment by smithfarm** [comment]
I ran it on a single-node cluster and got:

```
$ sesdev --log-file prd destroy --destroy-networks octopus
Are you sure you want to destroy the cluster? [y/N]: y
=== Running shell command ===
vagrant destroy admin --force
==> admin: Removing domain...
libvirt: Network Driver error : Network not found: no network with matching name 'octopus2'
```

I assume the destroyed domain from the single-node cluster was associated with the network called `octopus2` but I don't know why we're now getting `libvirt: Network Driver error : Network not found: no network with matching name 'octopus2'`...

Maybe just need to silence the error so it doesn't appear on the screen?

**Comment by smithfarm** [comment]
The log:

```
2020-03-09 17:28:27,551 [INFO] [seslib.tools] Running sync command (.): ['vagrant', 'box', 'list']
2020-03-09 17:28:28,367 [INFO] [seslib.tools] Running sync command (/home/smithfarm/.sesdev/octopus): ['vagrant', 'status']
2020-03-09 17:28:29,577 [INFO] [seslib.tools] Running async command (/home/smithfarm/.sesdev/octopus): ['vagrant', 'destroy', 'admin', '--force']
2020-03-09 17:28:32,083 [INFO] [seslib] Destroy network 'octopus2'
2020-03-09 17:28:32,084 [INFO] [seslib] Destroying network 'octopus2'
2020-03-09 17:28:32,084 [WARNING] [seslib] Unable to find network 'octopus2' for removal
2020-03-09 17:28:32,084 [WARNING] [seslib] Unable to destroy network 'octopus2' for deployment 'octopus'
```

File: seslib/__init__.py [hunk 2]
```diff
@@ -1030,6 +1088,16 @@ def destroy(self, log_handler):
             for image in images_to_remove:
                 self.box.remove_image(image)
 
+        for network in used_networks:
+            logger.info("Destroy network '%s'", network)
```
Comment (smithfarm) [inline seslib/__init__.py]: can this log message be dropped? It seems redundant - in the log I'm getting:

```
2020-03-09 17:28:32,083 [INFO] [seslib] Destroy network 'octopus2'
2020-03-09 17:28:32,084 [INFO] [seslib] Destroying network 'octopus2'
```

(the first is emitted here in `destroy` and the second immediately upon entering `destroy_network`)

File: seslib/__init__.py [hunk 3]
```diff
@@ -502,6 +529,26 @@ def remove_image(self, image_name):
     def remove_box(box_name):
         tools.run_sync(["vagrant", "box", "remove", box_name])
 
+    def destroy_network(self, name):
+        logger.info("Destroying network '%s'", name)
+        self.open_libvirt_connection()
+
+        try:
+            network = self.libvirt_conn.networkLookupByName(name)
```
Comment (smithfarm) [inline seslib/__init__.py]: I guess the error message `libvirt: Network Driver error : Network not found: no network with matching name 'octopus2'` is begin emitted by this line?

**Pull request merged**